app/src/ui/merge-branch/merge.tsx

```diff
@@ -283,13 +285,26 @@ export class Merge extends React.Component<IMergeProps, IMergeState> {
 
     const disabled = invalidBranchState || cannotMergeBranch
 
+    // the amount of characters to allow before we truncate was chosen arbitrarily
+    const currentBranchName = truncateWithEllipsis(
+      this.props.currentBranch.name,
+      40
+    )
     return (
       <Dialog
         id="merge"
-        title={__DARWIN__ ? 'Merge Branch' : 'Merge branch'}
         onDismissed={this.props.onDismissed}
         onSubmit={this.merge}
       >
+        <DialogHeader
+          title={
+            <div className="merge-dialog-header">
+              Merge into <b>{currentBranchName}</b>
+            </div>
+          }
+          dismissable={true}
+          onDismissed={this.props.onDismissed}
```

Comment:
I'm not a fan of how we've duplicated `onDismissed` here, and was hoping #5955 would help avoid this. I had a go at adding the support for `Dialog` to accept a `JSX.Element` but there's a bit that I'm not sure of (the `???` part).

```
diff --git a/app/src/ui/dialog/dialog.tsx b/app/src/ui/dialog/dialog.tsx
index 63162f312..9ff9cdee7 100644
--- a/app/src/ui/dialog/dialog.tsx
+++ b/app/src/ui/dialog/dialog.tsx
@@ -31,7 +31,7 @@ interface IDialogProps {
    * By omitting this consumers may use their own custom DialogHeader
    * for when the default component doesn't cut it.
    */
-  readonly title?: string
+  readonly title?: string | JSX.Element
 
   /**
    * Whether or not the dialog should be dismissable. A dismissable dialog
@@ -178,9 +178,11 @@ export class Dialog extends React.Component<IDialogProps, IDialogState> {
       this.setState({ titleId: undefined })
     }
 
-    if (this.props.title) {
+    if (this.props.title != null) {
+      // TODO: how can we get a string-like representation of a JSX element?
+      const id = typeof this.props.title === 'string' ? this.props.title : '???'
       this.setState({
-        titleId: createUniqueId(`Dialog_${this.props.id}_${this.props.title}`),
+        titleId: createUniqueId(`Dialog_${this.props.id}_${id}`),
       })
     }
   }
diff --git a/app/src/ui/merge-branch/merge.tsx b/app/src/ui/merge-branch/merge.tsx
index bd4239bc7..0d0ed27a3 100644
--- a/app/src/ui/merge-branch/merge.tsx
+++ b/app/src/ui/merge-branch/merge.tsx
@@ -19,7 +19,6 @@ import { MergeResultKind } from '../../models/merge'
 import { MergeStatusHeader } from '../history/merge-status-header'
 import { promiseWithMinimumTimeout } from '../../lib/promise'
 import { truncateWithEllipsis } from '../../lib/truncate-with-ellipsis'
-import { DialogHeader } from '../dialog/header'
 
 interface IMergeProps {
   readonly dispatcher: Dispatcher
@@ -293,18 +292,15 @@ export class Merge extends React.Component<IMergeProps, IMergeState> {
     return (
       <Dialog
         id="merge"
+        title={
+          <div className="merge-dialog-header">
+            Merge into <b>{currentBranchName}</b>
+          </div>
+        }
         onDismissed={this.props.onDismissed}
+        dismissable={true}
         onSubmit={this.merge}
       >
-        <DialogHeader
-          title={
-            <div className={'merge-dialog-header'}>
-              Merge into <b>{currentBranchName}</b>
-            </div>
-          }
-          dismissable={true}
-          onDismissed={this.props.onDismissed}
-        />
         <DialogContent>
           <BranchList
             allBranches={this.props.allBranches}
```

Reply:
Could we use the `key` property JSX elements are supposed to have?

Reply:
hey @shiftkey i think something like that makes sense and i'm happy to open that refactor in another PR to help with cases like these. i think we should merge this separately though.

Reply:
Moved to https://github.com/desktop/desktop/issues/6032